distributed/worker.py

```diff
@@ -1589,474 +1750,554 @@ def add_task(
+        start_finish = (start, finish_state)
+        func = self._transitions_table.get(start_finish)
 
-            self.waiting_for_data_count -= len(ts.waiting_for_data)
-            ts.waiting_for_data.clear()
-            if value is not None:
-                self.put_key_in_memory(ts, value)
-            self.send_task_state_to_scheduler(ts)
-        except Exception as e:
-            logger.exception(e)
-            if LOG_PDB:
-                import pdb
+        default_state = "released"
+        if func is not None:
+            a: tuple = func(ts, *args, stimulus_id=stimulus_id, **kwargs)
+            self._transition_counter += 1
+            recommendations, scheduler_msgs = a
+            self._notify_plugins("transition", ts.key, start, finish_state, **kwargs)
 
-                pdb.set_trace()
-            raise
-
-    def transition_ready_executing(self, ts):
-        try:
-            if self.validate:
-                assert not ts.waiting_for_data
-                assert ts.key not in self.data
-                assert ts.state in READY
-                assert ts.key not in self.ready
-                assert all(
-                    dep.key in self.data or dep.key in self.actors
-                    for dep in ts.dependencies
+        elif default_state not in start_finish:
+            try:
+                a: tuple = self._transition(ts, default_state, stimulus_id=stimulus_id)
+                a_recs, a_smsgs = a
+
+                recommendations.update(a_recs)
+                scheduler_msgs.extend(a_smsgs)
+                v = a_recs.get(ts, finish)
+                v_args = []
+                v_state = v
+                if isinstance(v, tuple):
+                    v_state, *v_args = finish
+                b: tuple = self._transition(
+                    ts, v_state, *v_args, stimulus_id=stimulus_id
                 )
+                b_recs, b_smsgs = b
+                recommendations.update(b_recs)
+                scheduler_msgs.extend(b_smsgs)
+            except (InvalidTransition, KeyError):
+                raise InvalidTransition(
+                    "Impossible transition from %r to %r for %s"
+                    % (*start_finish, ts.key)
+                ) from None
 
-            self.executing_count += 1
-            self.loop.add_callback(self.execute, ts.key)
-        except Exception as e:
-            logger.exception(e)
-            if LOG_PDB:
-                import pdb
+        else:
+            raise InvalidTransition(
+                "Impossible transition from %r to %r for %s" % (*start_finish, ts.key)
+            )
 
-                pdb.set_trace()
-            raise
+        self.log.append(
+            (
+                ts.key,
+                start,
+                ts.state,
+                {ts.key: new for ts, new in recommendations.items()},
+                stimulus_id,
+                time(),
+            )
+        )
+        return recommendations, scheduler_msgs
```

Comment:
Including the scheduler_msgs in this transition machine is likely overkill since most of our transitions do not include any messages. This is mostly in alignment to the scheduler